File: _daily_topic/tree/1104.py
from typing import List

class Solution:
    def pathInZigZagTree(self, label: int) -> List[int]:
        def getRow(label: int):
            row, start = 1, 1
            while (start * 2 <= label):
                row += 1
                start *= 2
            return row
        def getReverse(label: int, row: int):
            return (1 << row - 1) + (1 << row) - 1 - label
        row = getRow(label)
        if row & 1 == 1:
            label = getReverse(label)
        ans = []
        while row > 0:
            if row & 1 == 1:
                ans.append(getReverse(label, row))
            else:
                ans.append(label)
            row -= 1
            label >>= 1
        reversed(ans)
        return ans

File: _daily_topic/tree/1104.cpp
#include "header.h"
// 二叉树寻路
class Solution {
private:
    int getRow(int label) {
        int row = 1, rowStart = 1;
        while (rowStart * 2 <= label) {
            ++row;
            rowStart *= 2;
        }
        return row;
    }
    int getReverse(int label, int row) {
        return (1 << row - 1) + (1 << row) - 1 - label;
    }
public:
    vector<int> pathInZigZagTree(int label) {
        int row = getRow(label);
        if (row & 1 != 0) {
            label = getReverse(label, row);
        }
        vector<int> ans;
        while (row > 0) {
            if (row & 1 != 0) ans.push_back(getReverse(label, row));
            else ans.push_back(label);
            --row;
            label >>= 1;
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};